torch/testing/_internal/common_utils.py

```diff
@@ -803,204 +810,411 @@ def is_integral(self, dtype):
         elif isinstance(x, dict) and isinstance(y, dict):
             if isinstance(x, OrderedDict) and isinstance(y, OrderedDict):
                 self.assertEqual(x.items(), y.items(), atol=atol, rtol=rtol,
-                                 message=message, allow_inf=allow_inf, exact_dtype=exact_dtype)
+                                 message=message, exact_dtype=exact_dtype,
+                                 exact_device=exact_device)
             else:
                 self.assertEqual(set(x.keys()), set(y.keys()), atol=atol, rtol=rtol,
-                                 message=message, allow_inf=allow_inf, exact_dtype=exact_dtype)
+                                 message=message, exact_dtype=exact_dtype,
+                                 exact_device=exact_device)
                 key_list = list(x.keys())
                 self.assertEqual([x[k] for k in key_list],
                                  [y[k] for k in key_list],
                                  atol=atol, rtol=rtol, message=message,
-                                 allow_inf=allow_inf, exact_dtype=exact_dtype)
+                                 exact_dtype=exact_dtype, exact_device=exact_device)
         elif is_iterable(x) and is_iterable(y):
             super().assertEqual(len(x), len(y), message)
             for x_, y_ in zip(x, y):
                 self.assertEqual(x_, y_, atol=atol, rtol=rtol, message=message,
-                                 allow_inf=allow_inf, exact_dtype=exact_dtype)
+                                 exact_dtype=exact_dtype, exact_device=exact_device)
         elif isinstance(x, bool) and isinstance(y, bool):
-            super().assertEqual(x, y, message)
+            self.assertTrue(x == y, msg=message)
         elif isinstance(x, Number) and isinstance(y, Number):
-            if abs(x) == inf or abs(y) == inf:
-                if allow_inf:
-                    super().assertEqual(x, y, message)
-                else:
-                    self.fail("Expected finite numeric values - x={}, y={}".format(x, y))
+            # Short-circuits on identity
+            if x == y or (equal_nan and x != x and y != y):
                 return
-            super().assertLessEqual(abs(x - y), atol, message)
+
+            # Helper for comparing scalars and formatting debug info
+            def _scalar_compare_helper(message, s, x, y, rtol, atol):
+                diff = abs(x - y)
+                allowed_diff = atol + rtol * abs(y)
+                result = diff <= allowed_diff
+
+                if not result and message is None:
+                    message = ("Comparing " + s + "{0} and {1} gives a "
+                               "difference of {2}, but the allowed difference "
+                               "with rtol={3} and atol={4} is "
+                               "only {5}!").format(x, y, diff,
+                                                   rtol, atol, allowed_diff)
+
+                self.assertTrue(result, msg=message)
```

Comment:
isn't `result` a tensor here, with potentially many boolean values? does `assertTrue` assert that all values in the tensor are True?

Reply:
It's a bool. The comparison would be something like 4 < 5. No tensors involved here.